### PR TITLE
Begin implementation of build graph analysis

### PR DESCRIPTION
**Pull request by natemcmaster**
Adds a task that evaluations all solutions in universe and builds an in-memory object model about the projects, project references, and other properties.

FYI - most of this code has been copied from the code that was required to implement lineups.

cc @pranavkm

### CODE REVIEW

File: build/tasks/AnalyzeBuildGraph.cs [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+        /// New packages we are compiling. Used in the pin tool.
+        /// </summary>
+        [Output]
+        public ITaskItem[] PackagesProduced { get; set; }
+
+        /// <summary>
+        /// The order in which to build repositories
+        /// </summary>
+        [Output]
+        public ITaskItem[] RepositoryBuildOrder { get; set; }
```
Comment (JunTaoLuo) [inline build/tasks/AnalyzeBuildGraph.cs]: Where is this output set?

Comment (natemcmaster) [reply build/tasks/AnalyzeBuildGraph.cs]: Just pushed a change that includes it.

File: build/tasks/AnalyzeBuildGraph.cs [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+        public void Cancel()
+        {
+            _cts.Cancel();
+        }
+
+        public override bool Execute()
+        {
+            var factory = new SolutionInfoFactory(Log, BuildEngine5);
```
Comment (JunTaoLuo) [inline build/tasks/AnalyzeBuildGraph.cs]: Out of curiosity, what is BuildEngine5?

**Comment by natemcmaster** [comment]
Output so far:

image

Needs changes to KoreBuild. `IsPackable` isn't actually a good source of about which packages will be produced. We need access to the ProjectsToPack item group

File: build/tasks/AnalyzeBuildGraph.cs (hunk 2 above)
Comment (natemcmaster) [reply build/tasks/AnalyzeBuildGraph.cs]: Provides access to MSBuild's internal memory caching mechanisms.

**Comment by natemcmaster** [comment]
Depends on https://github.com/aspnet/BuildTools/pull/402 and https://github.com/aspnet/EntityFrameworkCore/pull/9834

**Comment by natemcmaster** [comment]
Package-output analysis is fixed, includes .Sources projects, and repo build order math is right too.

image

**Comment by natemcmaster** [comment]
More progress. I can detect inconsistencies in the package graph now. This should help us identify any issues in cascading versions.

image

File: build/tasks/AnalyzeBuildGraph.cs [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+        }
+
+        private void EnsureConsistentGraph(IEnumerable<PackageInfo> packages, IEnumerable<SolutionInfo> solutions)
+        {
+            // ensure versions cascade
+            var lookup = packages.ToDictionary(p => p.Id, p => p, StringComparer.OrdinalIgnoreCase);
+
+            var inconsistentVersions = new List<VersionMismatch>();
+
+            // holy crap, o^4
```
Comment (natemcmaster) [inline build/tasks/AnalyzeBuildGraph.cs]: Yes, i know. I'm going to refactor this later. I'm thinking projectmodel is basically a tree. We can probably use the visitor pattern and execute multiple visitors at once so we only have to pass over the tree once.

**Comment by natemcmaster** [comment]
Ok, I think this is ready now. Here are the kinds of errors our devs will see when they are preparing a patch. Output:

image

Usage:
```
./build.sh /t:Graph
```

File: build/tasks/GenerateLineup.cs [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+            foreach (var pkg in packages.OrderBy(i => i.Id))
+            {
+                var version = pkg.Version.ToString();
+                if (UseFloatingVersions && version.EndsWith(BuildNumber))
+                {
+                    version = version.Substring(0, version.Length - BuildNumber.Length) + "*";
+                }
+
+                var refType = "DotNetCliTool".Equals(pkg.Id, StringComparison.OrdinalIgnoreCase)
```
Comment (JunTaoLuo) [inline build/tasks/GenerateLineup.cs]: pkg.Id?

**Comment by natemcmaster** [comment]
So far, here is the list of the packages that are affected by this patch:
```xml
<Project>
  <ItemGroup>
    <PackageReference Update="Microsoft.AspNet.Identity.AspNetCoreCompat" Version="0.3.1-rtm-*" />
    <PackageReference Update="Microsoft.AspNetCore" Version="2.0.1-rtm-*" />
    <PackageReference Update="Microsoft.AspNetCore.All" Version="2.0.2-rtm-*" />
    <PackageReference Update="Microsoft.AspNetCore.Diagnostics" Version="2.0.1-rtm-*" />
    <PackageReference Update="Microsoft.AspNetCore.Diagnostics.Abstractions" Version="2.0.1-rtm-*" />
    <PackageReference Update="Microsoft.AspNetCore.Diagnostics.Elm" Version="0.3.1-rtm-*" />
    <PackageReference Update="Microsoft.AspNetCore.Diagnostics.EntityFrameworkCore" Version="2.0.1-rtm-*" />
    <PackageReference Update="Microsoft.AspNetCore.Diagnostics.Identity.Service" Version="2.0.1-rtm-*" />
    <PackageReference Update="Microsoft.AspNetCore.Identity" Version="2.0.1-rtm-*" />
    <PackageReference Update="Microsoft.AspNetCore.Identity.EntityFrameworkCore" Version="2.0.1-rtm-*" />
    <PackageReference Update="Microsoft.AspNetCore.Identity.Service" Version="2.0.1-rtm-*" />
    <PackageReference Update="Microsoft.AspNetCore.Identity.Service.Abstractions" Version="2.0.1-rtm-*" />
    <PackageReference Update="Microsoft.AspNetCore.Identity.Service.AzureKeyVault" Version="2.0.1-rtm-*" />
    <PackageReference Update="Microsoft.AspNetCore.Identity.Service.Core" Version="2.0.1-rtm-*" />
    <PackageReference Update="Microsoft.AspNetCore.Identity.Service.EntityFrameworkCore" Version="2.0.1-rtm-*" />
    <PackageReference Update="Microsoft.AspNetCore.Identity.Service.IntegratedWebClient" Version="2.0.1-rtm-*" />
    <PackageReference Update="Microsoft.AspNetCore.Identity.Service.Mvc" Version="2.0.1-rtm-*" />
    <PackageReference Update="Microsoft.AspNetCore.Identity.Service.Specification.Tests" Version="2.0.1-rtm-*" />
    <PackageReference Update="Microsoft.AspNetCore.Identity.Specification.Tests" Version="2.0.1-rtm-*" />
    <PackageReference Update="Microsoft.AspNetCore.MiddlewareAnalysis" Version="2.0.1-rtm-*" />
    <PackageReference Update="Microsoft.AspNetCore.Mvc" Version="2.0.1-rtm-*" />
    <PackageReference Update="Microsoft.AspNetCore.Mvc.Abstractions" Version="2.0.1-rtm-*" />
    <PackageReference Update="Microsoft.AspNetCore.Mvc.ApiExplorer" Version="2.0.1-rtm-*" />
    <PackageReference Update="Microsoft.AspNetCore.Mvc.Core" Version="2.0.1-rtm-*" />
    <PackageReference Update="Microsoft.AspNetCore.Mvc.Cors" Version="2.0.1-rtm-*" />
    <PackageReference Update="Microsoft.AspNetCore.Mvc.DataAnnotations" Version="2.0.1-rtm-*" />
    <PackageReference Update="Microsoft.AspNetCore.Mvc.Formatters.Json" Version="2.0.1-rtm-*" />
    <PackageReference Update="Microsoft.AspNetCore.Mvc.Formatters.Xml" Version="2.0.1-rtm-*" />
    <PackageReference Update="Microsoft.AspNetCore.Mvc.Localization" Version="2.0.1-rtm-*" />
    <PackageReference Update="Microsoft.AspNetCore.Mvc.Razor" Version="2.0.1-rtm-*" />
    <PackageReference Update="Microsoft.AspNetCore.Mvc.Razor.ViewCompilation" Version="2.0.1-rtm-*" />
    <PackageReference Update="Microsoft.AspNetCore.Mvc.RazorPages" Version="2.0.1-rtm-*" />
    <PackageReference Update="Microsoft.AspNetCore.Mvc.TagHelpers" Version="2.0.1-rtm-*" />
    <PackageReference Update="Microsoft.AspNetCore.Mvc.Testing" Version="2.0.1-alpha1-*" />
    <PackageReference Update="Microsoft.AspNetCore.Mvc.ViewFeatures" Version="2.0.1-rtm-*" />
    <PackageReference Update="Microsoft.AspNetCore.Mvc.WebApiCompatShim" Version="2.0.1-rtm-*" />
    <PackageReference Update="Microsoft.AspNetCore.NodeServices" Version="2.0.1-rtm-*" />
    <PackageReference Update="Microsoft.AspNetCore.NodeServices.Sockets" Version="2.0.1-rtm-*" />
    <PackageReference Update="Microsoft.AspNetCore.SpaServices" Version="2.0.1-rtm-*" />
    <PackageReference Update="Microsoft.EntityFrameworkCore" Version="2.0.1-rtm-*" />
    <PackageReference Update="Microsoft.EntityFrameworkCore.Design" Version="2.0.1-rtm-*" />
    <PackageReference Update="Microsoft.EntityFrameworkCore.InMemory" Version="2.0.1-rtm-*" />
    <PackageReference Update="Microsoft.EntityFrameworkCore.Relational" Version="2.0.1-rtm-*" />
    <PackageReference Update="Microsoft.EntityFrameworkCore.Relational.Design.Specification.Tests" Version="2.0.1-rtm-*" />
    <PackageReference Update="Microsoft.EntityFrameworkCore.Relational.Specification.Tests" Version="2.0.1-rtm-*" />
    <PackageReference Update="Microsoft.EntityFrameworkCore.Specification.Tests" Version="2.0.1-rtm-*" />
    <PackageReference Update="Microsoft.EntityFrameworkCore.Sqlite" Version="2.0.1-rtm-*" />
    <PackageReference Update="Microsoft.EntityFrameworkCore.Sqlite.Core" Version="2.0.1-rtm-*" />
    <PackageReference Update="Microsoft.EntityFrameworkCore.SqlServer" Version="2.0.1-rtm-*" />
    <PackageReference Update="Microsoft.EntityFrameworkCore.Tools" Version="2.0.1-rtm-*" />
    <PackageReference Update="Microsoft.EntityFrameworkCore.Tools.DotNet" Version="2.0.1-rtm-*" />
    <PackageReference Update="Microsoft.Extensions.Identity.Core" Version="2.0.1-rtm-*" />
    <PackageReference Update="Microsoft.Extensions.Identity.Stores" Version="2.0.1-rtm-*" />
    <PackageReference Update="Microsoft.VisualStudio.Web.CodeGeneration" Version="2.0.1-rtm-*" />
    <PackageReference Update="Microsoft.VisualStudio.Web.CodeGeneration.Contracts" Version="2.0.1-rtm-*" />
    <PackageReference Update="Microsoft.VisualStudio.Web.CodeGeneration.Core" Version="2.0.1-rtm-*" />
    <PackageReference Update="Microsoft.VisualStudio.Web.CodeGeneration.Design" Version="2.0.1-rtm-*" />
    <PackageReference Update="Microsoft.VisualStudio.Web.CodeGeneration.EntityFrameworkCore" Version="2.0.1-rtm-*" />
    <PackageReference Update="Microsoft.VisualStudio.Web.CodeGeneration.Templating" Version="2.0.1-rtm-*" />
    <PackageReference Update="Microsoft.VisualStudio.Web.CodeGeneration.Tools" Version="2.0.1-rtm-*" />
    <PackageReference Update="Microsoft.VisualStudio.Web.CodeGeneration.Utils" Version="2.0.1-rtm-*" />
    <PackageReference Update="Microsoft.VisualStudio.Web.CodeGenerators.Mvc" Version="2.0.1-rtm-*" />
  </ItemGroup>
</Project>
```

File: build/tasks/AnalyzeBuildGraph.cs [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+            var inconsistentVersions = new List<VersionMismatch>();
+            var reposThatShouldPatch = new HashSet<string>();
+
+            // holy crap, o^4
+            foreach (var sln in solutions)
+            foreach (var proj in sln.Projects)
+            foreach (var tfm in proj.Frameworks)
+            foreach (var dependency in tfm.Dependencies)
+            {
+                if (!buildPackageMap.TryGetValue(dependency.Key, out var package)) continue;
```
Comment (JunTaoLuo) [inline build/tasks/AnalyzeBuildGraph.cs]: When does this happen?

Comment (natemcmaster) [reply build/tasks/AnalyzeBuildGraph.cs]: I'll add a comment. This checks that the package is one that is actually being built by universe, and not an external or already-shipped dependency.

File: build/tasks/GenerateLineup.cs (hunk 4 above)
Comment (natemcmaster) [reply build/tasks/GenerateLineup.cs]: oops. copy posta err

File: build/repo.targets [hunk 6]
```diff
@@ -107,48 +105,69 @@
...
+
+    <Error Text="No solutions were found in '$(_CloneRepositoryRoot)'" Condition="@(Solution->Count()) == 0" />
   </Target>
 
-  <Target Name="_GenerateRestoreGraphSpecs" DependsOnTargets="_PrepareRestoreGraphSpecs" Inputs="@(GraphSpecInputs)" Outputs="@(GraphSpecOutputs)">
-    <MSBuild
-      Projects="@(Solution)"
-      Targets="GenerateRestoreGraphFile"
-      BuildInParallel="$(BuildInParallel)" />
+  <Target Name="Graph" DependsOnTargets="ResolveRepoInfo">
```
[the remark this answers is not in the document]
Comment (natemcmaster) [reply build/repo.targets]: ShowGraph?

File: build/tasks/AnalyzeBuildGraph.cs [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+        public ITaskItem[] Artifacts { get; set; }
+
+        // Artifacts that already shipped from repos
+        [Required]
+        public ITaskItem[] ShippedArtifacts { get; set; }
+
+        [Required]
+        public string Properties { get; set; }
+
+        public string StartGraphAt { get; set; }
```
[the remark this answers is not in the document]
Comment (natemcmaster) [reply build/tasks/AnalyzeBuildGraph.cs]: Can be remove for now. Would be nice to add back soon though

File: build/repo.targets (hunk 6 above)
[the remark this answers is not in the document]
Comment (JunTaoLuo) [reply build/repo.targets]: sure or ComputeGraph?

File: build/tasks/AnalyzeBuildGraph.cs (hunk 5 above)
Comment (JunTaoLuo) [reply build/tasks/AnalyzeBuildGraph.cs]: Hmm we need to probably maintain a list of external dependencies as well. I'd imagine we want to know if a corefx package we depended on updated and whether we need to bump our versions.

File: build/tasks/AnalyzeBuildGraph.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (JunTaoLuo) [reply build/tasks/AnalyzeBuildGraph.cs]: Computing the build graph doesn't take that long does it?

File: build/tasks/GenerateLineup.cs (hunk 4 above)
Comment (JunTaoLuo) [reply build/tasks/GenerateLineup.cs]: Ah, that makes more sense.

File: build/tasks/AnalyzeBuildGraph.cs (hunk 5 above)
Comment (natemcmaster) [reply build/tasks/AnalyzeBuildGraph.cs]: It's like you read my mind. The next thing I wanted to do was add this so we can also issue errors/warnings about undefined external dependencies

File: build/repo.targets (hunk 6 above)
[the remark this answers is not in the document]
Comment (natemcmaster) [reply build/repo.targets]: Lets go with ComputeGraph

File: build/tasks/AnalyzeBuildGraph.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (natemcmaster) [reply build/tasks/AnalyzeBuildGraph.cs]: about a minute or so

[the remark this answers is not in the document]
Comment (natemcmaster) [reply build/tasks/AnalyzeBuildGraph.cs]: This was copied from the CalculateBuildGraph task (which can be deleted now) and is used to allow people to start the build at Mvc and skip lower-level repos.

[the remark this answers is not in the document]
Comment (JunTaoLuo) [reply build/tasks/AnalyzeBuildGraph.cs]: That seems short enough that it isn't too terrible to rebuild the graph every time. But then again, no reason to remove it as I realized there's not that much code currently unused.

File: build/tasks/AnalyzeBuildGraph.cs [hunk 8]
```diff
@@ -0,0 +1,225 @@
...
+                .Where(p => !p.IsSymbolsArtifact)
+                .ToDictionary(p => p.PackageInfo.Id, p => p, StringComparer.OrdinalIgnoreCase);
+
+            // ensure versions cascade
+            var buildPackageMap = packages.ToDictionary(p => p.PackageInfo.Id, p => p, StringComparer.OrdinalIgnoreCase);
+
+            var inconsistentVersions = new List<VersionMismatch>();
+            var reposThatShouldPatch = new HashSet<string>();
+
+            // holy crap, o^4
```
Comment (pranavkm) [inline build/tasks/AnalyzeBuildGraph.cs]: remove

File: build/tasks/GenerateLineup.cs [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+            var doc = new XDocument(root);
+
+            var packages = new List<PackageInfo>();
+
+            foreach (var item in Artifacts)
+            {
+                var info = ArtifactInfo.Parse(item);
+                switch (info)
+                {
+                    case ArtifactInfo.Package pkg when (!pkg.IsSymbolsArtifact):
```
Comment (pranavkm) [inline build/tasks/GenerateLineup.cs]: Make this an `if`. Shouldn't need a switch for a single case.

File: build/tasks/ProjectModel/ProjectInfoFactory.cs [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+using Microsoft.Build.Construction;
+using Microsoft.Build.Evaluation;
+using Microsoft.Build.Execution;
+using NuGet.Frameworks;
+using RepoTasks.Utilities;
+using Microsoft.Build.Utilities;
+
+namespace RepoTasks.ProjectModel
+{
+    internal class ProjectInfoFactory
```
Comment (pranavkm) [inline build/tasks/ProjectModel/ProjectInfoFactory.cs]: Can't we just shell out to `GenerateRestoreGraphFile` and adapt it to our needs? We can pretty much infer most of these from the spec file, no?

Comment (natemcmaster) [reply build/tasks/ProjectModel/ProjectInfoFactory.cs]: The restore graph file was okay, but doesn't give us the complete story. Running a design-time build instead gives us the same information, and allows us to get even more, such as specific MSBuild properties like 'PackageType' and the location of obj/ and bin/.

Comment (pranavkm) [reply build/tasks/ProjectModel/ProjectInfoFactory.cs]: What do we need ouput paths for? Presumably all we need for lineups is package version information.

Comment (natemcmaster) [reply build/tasks/ProjectModel/ProjectInfoFactory.cs]: Not used yet. This code was copied from the dev branch of KoreBuild from the lineups task. I'm trying to keep it as similar as possible to reduce our pain when porting this work to dev.

**Comment by natemcmaster** [comment]
I believe this is ready to merge. I'm able to build a 2.0.2 patch locally now

File: build/tasks/ProjectModel/ProjectInfoFactory.cs (hunk 10 above)
Comment (pranavkm) [reply build/tasks/ProjectModel/ProjectInfoFactory.cs]: Fair enough

Comment (natemcmaster) [reply build/tasks/ProjectModel/ProjectInfoFactory.cs]: Also, it turns out that calculating the build graph this way is faster. It's about twice as fast as the previous method which generated a JSON file using /t:GenerateRestoreGraphFile

**Pull request merged**